PerimeterXModule/Internals/PxContext.cs

```diff
@@ -7,167 +7,257 @@
+		public Dictionary<string,string> OriginalTokens { get; set; }
+		public IPxCookie OriginalToken { get; set; }
+		public string OriginalTokenError { get; set; }
+		public string OriginalUUID { get; set; }
+		public object DecodedOriginalToken { get; set; }
+
+		public PxContext(HttpContext context, PxModuleConfigurationSection pxConfiguration)
+		{
+			ApplicationContext = context;
+
+			CookieOrigin = CookieOrigin.COOKIE;
+			PxCookies = new Dictionary<string, string>();
+			OriginalTokens = new Dictionary<string, string>();
+			S2SCallReason = "none";
+
+			// Get Headers
+
+			// if userAgentOverride is present override the default user-agent
+			string userAgentOverride = pxConfiguration.UserAgentOverride;
+			if (!string.IsNullOrEmpty(userAgentOverride))
+			{
+				UserAgent = context.Request.Headers[userAgentOverride];
+			}
+			// ua fallback
+			if (string.IsNullOrEmpty(UserAgent))
+			{
+				UserAgent = context.Request.Headers["user-agent"];
+			}
+
+			Headers = new List<RiskRequestHeader>();
+
+			foreach (string header in context.Request.Headers.Keys)
+			{
+				if (!pxConfiguration.SensitiveHeaders.Contains(header))
+				{
+					RiskRequestHeader riskHeader = new RiskRequestHeader
+					{
+						Name = header,
+						Value = context.Request.Headers.Get(header)
+					};
+					if (header.ToLower() == "user-agent")
+					{
+						riskHeader.Value = UserAgent;
+					}
+					Headers.Add(riskHeader);
+				}
+			}
+
+			// Handle Cookies
+			var contextCookie = context.Request.Cookies;
+			string mobileHeader = context.Request.Headers[PxConstants.MOBILE_HEADER];
+			// Check if X-PX-AUTHORIZATION exist
+			if (mobileHeader != null)
+			{
+				// Extract Original Tokens
+				CookieOrigin = CookieOrigin.HEADER;
+				string originalToken = context.Request.Headers[PxConstants.ORIGINAL_TOKEN];
+				if (!string.IsNullOrEmpty(originalToken))
+				{
+					string[] splitedOriginalToken = originalToken.Split(MOBILE_DELIMITER, 2);
+					string[] fallbackSplittedOriginalToken = originalToken.Split(MOBILE_DELIMITER);
+					if (!string.IsNullOrEmpty(splitedOriginalToken[0]) && Array.IndexOf(PxConstants.PX_TOKEN_PREFIX, splitedOriginalToken[0]) > -1)
+					{
+						string originlTokenKey = splitedOriginalToken[0].Equals(PxConstants.TOKEN_V3_PREFIX) ? PxConstants.COOKIE_V3_PREFIX : PxConstants.COOKIE_V1_PREFIX;
+						OriginalTokens.Add(originlTokenKey, splitedOriginalToken[1]);
+					}
+					// Fallback
+					else if (fallbackSplittedOriginalToken.Length == 4)
+					{
+						OriginalTokens.Add(PxConstants.COOKIE_V3_PREFIX, originalToken);
+					}
+					else if (fallbackSplittedOriginalToken.Length == 3)
+					{
+						OriginalTokens.Add(PxConstants.COOKIE_V1_PREFIX, originalToken);
+					}
+					else
+					{
+						OriginalTokens.Add(PxConstants.COOKIE_V3_PREFIX, originalToken);
+					}
+				}
+
+				// Extact Token
+				string[] splitedToken = mobileHeader.Split(MOBILE_DELIMITER, 2, StringSplitOptions.RemoveEmptyEntries);
```

Comment:
typo - `splittedToken`. 
BTW - I assume that we don't use `StringSplitOptions.RemoveEmptyEntries` when splitting `originalToken` on purpose?

Reply:
Why not do this logic (`mobileHeader.Split` etc) above the logic that has to do with `originalToken`? It would then correspond better with [your design doc](https://perimeterx.atlassian.net/wiki/spaces/PD/pages/120160898/Enforcer+-+Handling+X-PX-ORIGINGAL-TOKEN), and regardless it makes sense to have this logic there, right after you enter the `if (!string.IsNullOrEmpty(originalToken))`.